### PR TITLE
enable `on_load_checkpoint` for `datamodule` for all `trainer_fn`

### LINKED ISSUE #10229: call `on_load_checkpoint` for `LightningDataModule`
## Proposed refactoring or deprecation

call `on_load_checkpoint` for `LightningDataModule`.
currently it only works in case of fitting, but since datamodule is a part of training/evaluation process I think we should consider calling this hook on validate/test/predict calls.

### Motivation

If I implement the complete data logic inside `LightningModule` and want to reload any data related states, it will work, but if I use a datamodule it won't.

Is there any reason why this was ignored?

### Pitch

Simple fix. Move it up and update/add tests.
https://github.com/PyTorchLightning/pytorch-lightning/blob/dbe1662dc38d5217328ab459743b1113869a628c/pytorch_lightning/trainer/trainer.py#L1017-L1018

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10229
In the past, datamodule states were loaded only with `.fit` while resuming the training using `Trainer(resume_from_checkpoint=...)`. But now since we allow checkpoint for other trainer methods, and datamodule contains all the data-related info, it seems reasonable to load its states as well during `.test/predict/validate`. The PR changes this behavior.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_datamodules.py [hunk 1]
```diff
@@ -24,6 +24,8 @@
 
 from pytorch_lightning import LightningDataModule, Trainer
 from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.trainer.states import TrainerFn
+from pytorch_lightning.trainer.supporters import CombinedLoader
```
Comment (rohitgr7) [inline tests/core/test_datamodules.py:28]:
```suggestion
```

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by SeanNaren — APPROVED** [review]
